pybdv/bdv_datasets.py

```diff
@@ -0,0 +1,213 @@
+        downscale_mode=downscale_mode,
+        n_threads=n_threads
+    )
+
+    # Now, we just need to fetch the scaled data and put it to the proper positions
+    for scale_id, scale in enumerate(scales):
+
+        # Position in the current scale
+        pos_in_scale = (target_pos / scale).astype(int)
+        shp_in_scale = (target_shape / scale).astype(int)
+
+        # Write the data to each scale
+        with open_file(tmp_bdv, mode='r') as f:
+            scaled_vol = f[f'setup0/timepoint0/s{scale_id}'][:]
+            # assert list(scaled_vol.shape) == shp_in_scale.tolist()
+
+        with open_file(data_path, mode='a') as f:
+            f[f'setup{setup_id}/timepoint{timepoint}/s{scale_id}'][
+                pos_in_scale[0]: pos_in_scale[0] + shp_in_scale[0],
+                pos_in_scale[1]: pos_in_scale[1] + shp_in_scale[1],
+                pos_in_scale[2]: pos_in_scale[2] + shp_in_scale[2]
+            ] = scaled_vol
+
+    # Delete the temporary bdv file
+    rmtree(tmp_bdv)
+    os.remove(tmp_name + '.xml')
+
+
+class BdvDataset:
+    """
+    The basic BDV dataset to which volumes can be written using numpy nomenclature.
+
+    The data is included into each of the down-sampling layers accordingly
+    The full resolution area is padded, if necessary, to avoid sub-pixel locations in the down-sampling layers
+    """
+
+    def __init__(self, path, timepoint, setup_id, n_threads=1, verbose=False):
+
+        self._path = path
+        self._timepoint = timepoint
+        self._setup_id = setup_id
+        self._n_threads = n_threads
+        self._verbose = verbose
+
+    def _add_to_volume(self, position, volume):
+
+        # Determine required scales
+        with open(os.path.join(self._path, 'setup{}'.format(self._setup_id), 'attributes.json')) as f:
+            scales = np.array(json.load(f)['downsamplingFactors'])
+
+        # Determine full dataset shape
+        with open_file(self._path, mode='r') as f:
+            full_shape = f[f'setup{self._setup_id}/timepoint{self._timepoint}/s0'].shape
+
+        if self._verbose:
+            print(f'scales = {scales}')
+            print(f'full_shape = {full_shape}')
+
+        # Check for out of bounds (and fix it if not)
+        position, volume = _check_for_out_of_bounds(position, volume, full_shape, verbose=self._verbose)
+
+        # Check if volume and position properly scale to the final scale level (and fix it if not)
+        max_scale = scales[-1]
+        target_vol, target_pos, target_shape = _check_shape_and_position_scaling(
+            max_scale, position, volume,
+            self._path, self._setup_id, self._timepoint,
+            verbose=self._verbose)
+
+        # Scale volume and write to target dataset
+        _scale_and_add_to_dataset(self._path, self._setup_id, self._timepoint,
+                                  target_pos, target_vol, target_shape,
+                                  scales, 'interpolate',
```

Comment:
We shouldn't hard-code the interpolation mode, but rather pass it as a parameter to the class, so that this works for label volumes or masks as well.